network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -0,0 +1,151 @@
+	authorizedRolesMap[cborcodec.CodeApprovalRequest] = flow.RoleList{flow.RoleConsensus}
+	authorizedRolesMap[cborcodec.CodeApprovalResponse] = flow.RoleList{flow.RoleVerification}
+
+	// generic entity exchange engines
+	authorizedRolesMap[cborcodec.CodeEntityRequest] = flow.RoleList{flow.RoleAccess, flow.RoleConsensus, flow.RoleCollection} // only staked access nodes
+	authorizedRolesMap[cborcodec.CodeEntityResponse] = flow.RoleList{flow.RoleCollection, flow.RoleExecution}
+
+	// testing
+	authorizedRolesMap[cborcodec.CodeEcho] = flow.Roles()
+
+	// dkg
+	authorizedRolesMap[cborcodec.CodeDKGMessage] = flow.RoleList{flow.RoleConsensus} // sn nodes for next epoch
+}
+
+// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
+// is part of the authorized roles list for the type of message being sent. A node is considered
+// to be authorized to send a message if all of the following are true.
+// 1. The message type is a known message type (initialized in the authorizedRolesMap).
+// 2. The authorized roles list for the message type contains the senders role.
+// 3. The node has a weight > 0 and is not ejected
+func AuthorizedSenderValidator(log zerolog.Logger, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
+	log = log.With().
+		Str("component", "authorized_sender_validator").
+		Logger()
+
+	return func(ctx context.Context, from peer.ID, msg *message.Message) pubsub.ValidationResult {
+		identity, ok := getIdentity(from)
+		if !ok {
+			log.Warn().Str("peer_id", from.String()).Msg("could not get identity of sender")
+			return pubsub.ValidationReject
+		}
+
+		msgType := msg.Payload[0]
+
+		if err := isAuthorizedSender(identity, msgType); err != nil {
+			log.Warn().
+				Err(err).
+				Str("peer_id", from.String()).
+				Str("role", identity.Role.String()).
+				Uint8("message_type", msgType).
+				Msg("rejecting message")
+
+			return pubsub.ValidationReject
+		}
+
+		return pubsub.ValidationAccept
+	}
+}
+
+func isAuthorizedSender(identity *flow.Identity, msgType uint8) error {
+	err := isAuthorizedNodeRole(identity.Role, msgType)
+	if err != nil {
+		return err
+	}
+
+	err = isActiveNode(identity.NodeID, identity.Weight, identity.Ejected)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// isAuthorizedNodeRole checks if a role is authorized to send message type
+func isAuthorizedNodeRole(role flow.Role, msgType uint8) error {
+	roleList, ok := authorizedRolesMap[msgType]
+	if !ok {
+		return fmt.Errorf("unknown message type does not match any code from the cbor codec")
+	}
+
+	if !roleList.Contains(role) {
+		return fmt.Errorf("sender is not authorized to send this message type")
+	}
+
+	return nil
+}
+
+// isActiveNode checks that the node has a weight > 0 and is not ejected
+func isActiveNode(nodeID flow.Identifier, weight uint64, ejected bool) error {
+	if weight <= 0 {
+		return fmt.Errorf("node %s has an invalid weight of %d is not an active node", nodeID, weight)
```

Comment:
⚠️  Nodes with zero weight need to be able to participate in network communications, for a limited number of messages. We should only discard messages at the network layer on the sole basis of a node's identity, if a node is ejected. In particular, this is needed for the DKG and for block synchronization for new nodes. More details [here](https://www.notion.so/dapperlabs/Epoch-Networking-0c7b7a6ecd904a9fb8c023069488fe51).

In the interest of keeping with the approach here, my suggestion would be to implement these restrictions alongside the role-based restrictions added in this PR. For each message code, in addition to an allow-list of sender roles, we can include a tag for whether the message can be sent be a zero-weight node.

Reply:
Got it, so just to confirm: 
- It is safe to discard messages from ejected nodes ([via this link](https://www.notion.so/dapperlabs/Epoch-Networking-0c7b7a6ecd904a9fb8c023069488fe51)
- Certain messages must be allowed for nodes with 0 weight ?

Do you know where I can find this list of limited messages wrt the second point ?

Reply:
> Do you know where I can find this list of limited messages wrt the second point ?

I don't think we have a list anywhere, but it is:
* everything needed to synchronize chain state (sync requests/responses, block responses, pushed blocks)
* DKG messages

In terms of channels:
* `consensus-committee`
* `sync-committee`
* `dkg-committee`
* `push-blocks`

Reply:
```
	// consensus
	authorizedRolesMap[cborcodec.CodeBlockProposal] = flow.RoleList{flow.RoleConsensus}
	authorizedRolesMap[cborcodec.CodeBlockVote] = flow.RoleList{flow.RoleConsensus}

	// protocol state sync
	authorizedRolesMap[cborcodec.CodeSyncRequest] = flow.Roles()
	authorizedRolesMap[cborcodec.CodeSyncResponse] = flow.Roles()
	authorizedRolesMap[cborcodec.CodeRangeRequest] = flow.Roles()
	authorizedRolesMap[cborcodec.CodeBatchRequest] = flow.Roles()
	authorizedRolesMap[cborcodec.CodeBlockResponse] = flow.Roles()
        authorizedRolesMap[cborcodec.CodeDKGMessage] = flow.RoleList{flow.RoleConsensus}
```

This should be the full list, can you confirm ?

Reply:
That looks correct. Only change is, we can remove `CodeBlockVote` (zero-weight nodes aren't voting).

Reply:
I noted in a separate comment, but I discussed with with Alex, and we should skip the weight checks in the topic validator since it's not practical to know the specific view/block a message relates to outside of the application layer. We can still do the ejected checks since those take effect immediately.

Reply:
removed 0 weight checks and test so we don't need this list anymore https://github.com/onflow/flow-go/pull/2317/commits/079042c5ef705456cae43752bd375f29afdec63c